### ISSUE TITLE
Decorators have the same line number as the functions they decorate

### ISSUE THREAD

**Issue by EgorBu** [bblfsh/python-driver#63]
Hi,
I tested https://github.com/EgorBu/uast_playground at a new file
`python3 -m uast_playground repo2id_str -r test_data/exp2`

![screenshot from 2017-08-10 11-02-10](https://user-images.githubusercontent.com/2606702/29163053-8c624b94-7dbb-11e7-9e30-60cb63135e7f.png)

1) wrong start and end positions of tokens (token has more than 1 line, overlaps, wrong length, etc)
3) decorators and function definitions have the same start line

**Comment by juanjux** [comment]
The token of a function is the name, but the start and ending positions are the ones of the function node itself with body and all, not the token. This is the same for all nodes. The current UAST specification doesn't give a position for the tokens except when the token match the length of the node (variable names, etc.)

**Comment by EgorBu** [comment]
but at least start position should be the same as in code - don't you think so?

**Comment by juanjux** [comment]
No, because if `def somefunc` or `class SomeClass` the start position is the position of `def` or `class`, not of `somefunc`.

**Comment by EgorBu** [comment]
so it will be position of def, class, self?
what about a.b? is the position of b == position of a?

**Comment by juanjux** [comment]
The code `a.b` produces this native AST. As you can see the `col_offsets` are different:

```json
{'ast_type': 'Module',
 'body': [{'ast_type': 'Expr',
           'col_offset': 1,
           'end_col_offset': 3,
           'end_lineno': 1,
           'lineno': 1,
           'value': {'ast_type': 'Attribute',
                     'attr': 'b',
                     'col_offset': 3,
                     'ctx': 'Load',
                     'end_col_offset': 3,
                     'end_lineno': 1,
                     'lineno': 1,
                     'value': {'ast_type': 'Name',
                               'col_offset': 1,
                               'ctx': 'Load',
                               'end_col_offset': 1,
                               'end_lineno': 1,
                               'id': 'a',
                               'lineno': 1}}}],
 'col_offset': 1,
 'end_col_offset': 3,
 'end_lineno': 1,
 'lineno': 1}
```

**Comment by EgorBu** [comment]
can you explain about line number, col number, col offset, end col offset? Do we have it somewhere in bblfsh docs?

**Comment by juanjux** [comment]
This is an example of the native AST, you don't have to worry about it, it would be translated to StartPosition and EndPosition in the UAST (with some corrections).

**Comment by EgorBu** [comment]
```
for key, val in some_dict.items():
...
```
`key` & `val` has the same positions according to image.
Is it because of start position right after `for`?

**Comment by EgorBu** [comment]
In Repo2nBOW class definition:
```
MODEL_CLASS = NBOW
```
and it was converted to
```
MODEL_CLANBOW0m = NBOW
```
`0m` from color definition `WHITE = "\033[0m"`, so it looks like that in this case position somehow is shifted.

**Comment by EgorBu** [comment]
btw - what is the right way to estimate a position of the token in code?

**Comment by juanjux** [comment]
Code:

```python
class C():
    MODEL_CLANBOW = NBOW
```

Gives me an UAST with the columns 5 and 17 for `MODEL_CLANBOW` which is Vim is not lying to me are correct.

**Comment by juanjux** [comment]
Code:

```python
for key, val in some_dict.items(): pass
```

Gives me this (reduced and sorted) UAST:

```python
                     {
                        "Token" : "key",
                        "StartPosition" : {
                           "Line" : 1,
                           "Col" : 5,
                           "Offset" : 4
                        },
                        "EndPosition" : {
                           "Line" : 1,
                           "Col" : 7,
                           "Offset" : 6
                        }
                     },
                     {
                        "Token" : "val",
                        "StartPosition" : {
                           "Line" : 1,
                           "Col" : 10,
                           "Offset" : 9
                        },
                        "EndPosition" : {
                           "Line" : 1,
                           "Col" : 12,
                           "Offset" : 11,
                        },
                     }
```

Please check the UAST produced.

**Comment by EgorBu** [comment]
So initial code:
```
class Repo2nBOW(Repo2Base):
    """
    Implements the step repository -> :class:`ast2vec.nbow.NBOW`.
    """
    MODEL_CLASS = NBOW
```
and after processing:
```
class Repo2nBOW(Repo2Base):!Wrong length of token 'Repo2nBOW' (expected 22 - 7 + 1 == 9)!!token 'Repo2nBOW' start & end line: 14, 18!
    """
    Implements the step repository -> :class:`ast2vec.nbow.NBOW`.
    """
    MODEL_CLANBOW0m = NBOW
```

**Comment by juanjux** [comment]
@EgorBu I don't know where that `0m` comes from but in my test the positions in the UAST are correct - check it on your own directly without the uast_playground and tell me if you have different positions in the UAST.

**Comment by juanjux** [comment]
I've tested that the decorators are correct in their line numbers. It's the `FunctionDef` parent object that starts in the previous line... but this is correct at least with regard to the native Python AST since the decorators are part of the function AST, so for the language the function starts where the decorator (instead of in the `def` where undecorated functions start).

**Comment by EgorBu** [comment]
It will be really nice to mention it in the documentation.
Because these details are not obvious and can make a lot of people confused.

**Comment by juanjux** [comment]
Reopening since the documentation clearly says that the `StartPosition` of a node with a Token is the start position of the Token.

**Comment by juanjux** [comment]
Should be fixed in v0.7.1 (update the client-python too), thanks for the report!
